lsp/src/fiber_stream.mli

```diff
@@ -0,0 +1,40 @@
+open Import
+
+module In : sig
+  type 'a t
+
+  val create : (unit -> 'a option Fiber.t) -> 'a t
+
+  val empty : unit -> 'a t
+
+  val of_list : 'a list -> 'a t
+
+  val read : 'a t -> 'a option Fiber.t
+
+  val map : 'a t -> f:('a -> 'b) -> 'b t
+
+  val filter : 'a t -> f:('a -> bool) -> 'a t
+
+  val filter_map : 'a t -> f:('a -> 'b option) -> 'b t
+end
+
+module Out : sig
+  type 'a t
+
+  val create : ('a option -> unit Fiber.t) -> 'a t
+
+  val write : 'a t -> 'a option -> unit Fiber.t
+
+  val of_ref : 'a list ref -> 'a t
+
+  val null : unit -> 'a t
+end
+
+(** [connect i o] read from [i] and write them to [o]. Close [o] when [i] is
```

Comment:
```suggestion
(** [connect i o] reads from [i] and writes to [o]. Closes [o] when [i] is
```
More importantly, if I look at the corresponding implementation, I don't understand how the channel gets closed on exhaustion ?
```ocaml
let connect i o =
  let open Fiber.O in
  let rec go () =
    let* a = In.read i in
    let* () = Out.write o a in
    go ()
  in
  go ()
```
I might be missing something here

Reply:
When `i` is closed, reading it will return `None`. Once we read that `None`, we will use write it to `o` which will close it.

Reply:
But, looking at `In` and `Out` implementation, `Out.write` and `In.read` have the following implementations:
```ocaml
(* Out *)
let write f x = f x

(* In *)
let read t = t ()
```
I don't understand how anything gets "closed", and therefore how the loop ever stops ?
For example, in this case:
```ocaml
let fiber = connect (In.of_list []) (Out.of_ref (ref []))
```
Isn't this just going to be an infinite loop ?

Reply:
Oops, I just noticed the problem and pushed a fix. We should add 

Just to clarify, a closed `In` stream is one that returned `None`. Any subsequent reads should return `None` as well. A closed `Out` stream is one that received a `None`. Writing more elements after `None` should be a no-op (perhaps it should even be an error - I'll check haskell's implementation).